### DISCUSSION TITLE
ServerRequestContext.with change with ServerRequestContext.currentRequest in micronaut 4

### DISCUSSION THREAD

**Discussion by ethanhere** [micronaut-projects/micronaut-core#10681] [Q&A] [answered]
I am not sure if this is related to https://github.com/micronaut-projects/micronaut-core/discussions/9606:

I am upgrading my application from micronaut 3 to 4 and facing a problem related to ServerRequestContext:
@MicronautTest

    @BeforeEach
    public void setup() {
...
        ServerRequestContext.with(this.activeRequest);
    }

    @Test
    Public void test() {
...
        final HttpRequest<Object> request = ServerRequestContext.currentRequest()
                .orElseThrow(new Exception());
..

}	

With micronaut 3, we are getting this.activeRequest object return from ServerRequestContext.currentRequest() call.

However, with micronaut 4, I don't know how to make that happen. Simply change the setup to
ServerRequestContext.with(this.activeRequest, () -> {} );
seems not working.

Can you suggest a solution?

**Comment by sdelamo** [comment]
Have you tried adding the io.micronaut:context-propagation dependency? https://docs.micronaut.io/4.3.12/guide/#reactorContextPropagation 

Do yo have a sample app which reproduces the issue?

**Reply by ethanhere** [reply]
@sdelamo I don't fully understand io.micronaut:context-propagation dependency's role here.

Before I work on a sample app, my real application code is described in the original description: in a @MicronautTest 's set up, we call ServerRequestContext.with(this.activeRequest); In the @Test method, it calls ServerRequestContext.currentRequest() and expects the this.activeRequest returned. That worked with micronaut 3.

In micronaut 4, we want the same. So there are 2 problems:
1. Interface change: micronaut 4 no longer has ServerRequestContext.with(HttpRequest). What should I change to?
ServerRequestContext.with(HttpRequest, () -> {}) ?
3. What should I do, in order to make ServerRequestContext.currentRequest() returns this.activeRequest object ?

**Reply by sdelamo** [reply]
do you have a sample app which reproduces what you describe with Micronaut Framework 3?

**Reply by ethanhere** [reply]
I do. However the app is my employer's proprietary software. As I explained, I just want to ServerRequestContext.currentRequest() to return a certain object in a test. With micronaut 3 we achieve that by ServerRequestContext.with(HttpRequest) in setup method. With Micronaut 4, can you suggest an approach we can do that?

**Reply by sdelamo** [reply]
@dstepanov do you know what changed in ServerRequestContext and how to achieve what @ethanhere is describing. 

@ethanhere if you can create a small sample of what you currently do ni Micronaut Framework 3 that. it is helpful to add it to the test suite.

**Reply by dstepanov** [reply]
@ethanhere Please provide an example that doesn't work; I don't understand what you are trying to do.

**Reply by ethanhere** [reply]
@dstepanov  @sdelamo :  Here is the example code attached. 
[ServerRequestContextTest.txt](https://github.com/micronaut-projects/micronaut-core/files/14907867/ServerRequestContextTest.txt)

I have to change the .java file extension to be .txt in order to upload...

The problem is simple, as in the file, I have comment:
        // micronaut 3 code:
        //ServerRequestContext.with(this.activeRequest);

With micronaut 3, 
`ServerRequestContext.with(this.activeRequest);`
will compile, and it will make line 40
ServerRequestContext.currentRequest()
returns an optinal of this.activeRequest.

With micronaut 4,
I changed
`ServerRequestContext.with(this.activeRequest);`
to be
`ServerRequestContext.with(this.activeRequest, ()->{System.out.println ("ServerRequestContext.with null");});`

that nolonger makes ServerRequestContext.currentRequest() to return this.activeRequest object.

I wonder what would be replacement syntax in micronaut 4 in order for ServerRequestContext.currentRequest() to return this.activeRequest object ?

**Reply by dstepanov** [reply, marked as answer]
You can do `PropagatedContext.getOrEmpty().plus(new ServerHttpRequestContext(request)).propagate()` and call close to cleanup. This might change in the next major version.

**Reply by ethanhere** [reply]
can you be more specific about "call close", can you give a sample code?

**Reply by dstepanov** [reply]
The method will return a `Scope` and calling `close` will reset the thread context.

**Reply by ethanhere** [reply]
thanks @dstepanov
